site/content/blog/2019-09-29-avoiding-flaky-tests.md

```diff
@@ -0,0 +1,202 @@
+
+In addition to a number of [length and character checks](https://github.com/mattermost/mattermost-server/blob/e5ba0a0a1808660d4885ef04888da437c2a5b6a0/model/team.go#L137-L196), the Mattermost server [rejects certain prefixes](https://github.com/mattermost/mattermost-server/blob/969c032a1e9cb2ce0ef2c59143c071dbf78cdcb1/model/utils.go#L360) to avoid team names like `signup` or `LOGIN21`.
+
+When unit testing, we often rely on `model.NewId` to generate random, UUID version 4 Guid values for testing:
+
+```go
+	t1 := model.Team{}
+	t1.DisplayName = "Name"
+	t1.Name = model.NewId()
+	t1.Email = MakeEmail()
+	t1.Type = model.TEAM_OPEN
+	_, err := ss.Team().Save(&t1)
+	require.Nil(t, err)
+```
+
+Every so often, a test would fail claiming that [the team name was invalid](https://mattermost.atlassian.net/browse/MM-17053). It turned out that, from to time, some of those reserved prefixes would be randomly selected for the start of a `model.NewId`, e.g. `APICE0E9143C...`. Random? Yes! Invalid? Yes! Expected? No.
+
+The [fix](https://github.com/mattermost/mattermost-server/pull/11705) was simple, and replicated the pattern used in most other tests to prefix the team name:
+
+```go
+	t1 := model.Team{}
+	t1.DisplayName = "Name"
+	t1.Name = "zz" + model.NewId()
+	t1.Email = MakeEmail()
+	t1.Type = model.TEAM_OPEN
+	_, err := ss.Team().Save(&t1)
+	require.Nil(t, err)
+```
+
+### 6) Don't assume the test is actually flaky!
+
+One danger of having flaky tests is conditioning developers to believe that **any** unexpected test failure is likely to be flaky. As an example, we had received [multiple reports](https://mattermost.atlassian.net/browse/MM-16479) of a flaky unit test concerning the `FileWillBeUploaded` plugin hook.  Much effort was spent statically analyzing the code and attempting to reproduce locally. Even with the test repeated 100x, it always succeeded locally. Given the ongoing impact to developers with unrelated changes, we were contemplating disabling the test altogether.
+
+Fastforward a few weeks, and a community member [reported](https://community-daily.mattermost.com/core/pl/oficg6nb9bb6igzz87fpumgxuo) a mysterious issue: sometimes files processed by a plugin using `FileWillBeUploaded` wouldn't successfully update. One of our newest staff members took a look, and immediately [spotted](https://community-daily.mattermost.com/core/pl/5p9fjcqk3tdw7brjpfm1f3poza) a race condition in the related code. The [resulting fix](https://github.com/mattermost/mattermost-server/pull/12249) adressed both the community member's issue and our longstanding "flaky" test.
+
+## Conclusion
+
+To summarize from our own observations:
+
+* File a ticket with as much detail as possible!
+* Don't use `time.Sleep`? Do use `time.Sleep`? Know when to use `time.Sleep` responsibly.
+* Know your database semantics.
+* Expect anything from the unexpected and plan accordingly.
+* Every flaky test is a bug. Don't assume it's always a bug in the test.
```

Comment:
One item I want is: "Don't rely on a stable connection to the internet". We have a couple of tests that fetch data from the internet and are flaky because of this. (E.g. https://mattermost.atlassian.net/browse/MM-18628) Do you think this point is valid and should be added?

Reply:
ah yes. Good point I don't think the tests should be fetching data over the internet anyway 🤷‍♂. That should be solved in another way, setting up own test server, reaching out to local docker instance, etc.